src/qibo/tests/test_hamiltonians.py

```diff
@@ -1,49 +1,319 @@
-    H = getattr(hamiltonians, model)(**kwargs)
-    matrix = K.to_numpy(H.matrix).flatten().real
-    assert_regression_fixture(matrix, filename)
-
-
-@pytest.mark.parametrize("nqubits", [3, 4])
-@pytest.mark.parametrize("dense,calcterms", [(True, False), (False, False), (False, True)])
-def test_maxcut(nqubits, dense, calcterms):
-    size = 2 ** nqubits
-    ham = np.zeros(shape=(size, size), dtype=np.complex128)
-    for i in range(nqubits):
-        for j in range(nqubits):
-            h = np.eye(1)
-            for k in range(nqubits):
-                if (k == i) ^ (k == j):
-                    h = np.kron(h, matrices.Z)
-                else:
-                    h = np.kron(h, matrices.I)
-            M = np.eye(2**nqubits) - h
-            ham += M
-    target_ham = K.cast(- ham / 2)
-    final_ham = hamiltonians.MaxCut(nqubits, dense)
-    if (not dense) and calcterms:
-        _ = final_ham.terms
-    K.assert_allclose(final_ham.matrix, target_ham)
+from qibo import hamiltonians
+from qibo.tests.utils import random_complex, random_sparse_matrix
+
+#TODO: check if using .data is a good way to compare sparse matrices
+
+def test_hamiltonian_init(backend):
+    with pytest.raises(TypeError):
+        H = hamiltonians.Hamiltonian(2, "test")
+    with pytest.raises(ValueError):
+        H1 = hamiltonians.Hamiltonian(-2, np.eye(4))
+    with pytest.raises(RuntimeError):
+        H2 = hamiltonians.Hamiltonian(np.eye(2), np.eye(4))
+    with pytest.raises(ValueError):
+        H3 = hamiltonians.Hamiltonian(4, np.eye(10))
+
+@pytest.mark.parametrize("dtype", [np.int, np.float, np.complex, np.int32,
+                                   np.int64, np.float32, np.float64,
+                                   np.complex64, np.complex128])
+@pytest.mark.parametrize("sparse_type", [None, "coo", "csr", "csc", "dia"])
+def test_hamiltonian_algebraic_operations(backend, dtype, sparse_type):
+    """Test basic hamiltonian overloading."""
+    def transformation_a(a, b):
+        c1 = dtype(0.1)
+        return a + c1 * b
+    def transformation_b(a, b):
+        c1 = dtype(2)
+        c2 = dtype(3.5)
+        return c1 * a - b * c2
+    def transformation_c(a, b, use_eye=False):
+        c1 = dtype(4.5)
+        if use_eye:
+            return a + c1 * np.eye(a.shape[0]) - b
+        else:
+            return a + c1 - b
+    def transformation_d(a, b, use_eye=False):
+        c1 = dtype(10.5)
+        c2 = dtype(2)
+        if use_eye:
+            return c1 * np.eye(a.shape[0]) - a + c2 * b
+        else:
+            return c1 - a + c2 * b
+
+    if sparse_type is None:
+        H1 = hamiltonians.XXZ(nqubits=2, delta=0.5)
+        H2 = hamiltonians.XXZ(nqubits=2, delta=1)
+        mH1, mH2 = backend.to_numpy(H1.matrix), backend.to_numpy(H2.matrix)
+    else:
+        mH1 = random_sparse_matrix(backend, 64, sparse_type=sparse_type)
+        mH2 = random_sparse_matrix(backend, 64, sparse_type=sparse_type)
+        H1 = hamiltonians.Hamiltonian(6, mH1)
```

Comment:
For example this should be:
```suggestion
        H1 = hamiltonians.Hamiltonian(6, mH1, backend=backend)
```